### ISSUE TITLE
Type Narrowing

### ISSUE THREAD

**Issue by mehdigmira** [microsoft/pyright#5694]
**Describe the bug**
Not 100% this is a bug or intended, but type narrowing doesn't work in the example below (tested with pyright 1.1.315 and pyright 1.1.321)

**Code or Screenshots**

```python
# pyright: strict


from typing import ClassVar


class A:
    ...


class B:
    b: ClassVar[int] = 1


class C(B):
    ...


def f(x: None | type[B] | type[C]):
    if x is not None:
        # x is narrowed to type[B] | type[C]
        return x.b


def g(x: type[A] | type[B] | type[C]):
    if x is not A:
        # x is not narrowed
        return x.b


```

**Comment by hmc-cs-mdrissi** [comment]
```python
class D(A):
  ...
```

If x is D then `x is not A`, but is still of type[A]. x: type[A] does not mean that x is exactly type A, but it can be any of it's subtypes. Like x: float does not mean type(x) is float, but type(x) is a subtype of float.

So I think this is expected behavior.

**Comment by mehdigmira** [comment]
thanks, closing this. Narrowing works wells when using issubclass(x, A) instead. Was confused because this error popped up when upgrading to 1.1.321 (I don't know why it wasn't flagged before)

**Comment by erictraut** [comment]
If you mark class `A` as `@final`, then type narrowing can be safely applied in this case because pyright can then assume that no subclasses of `A` exist.
